google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java

```diff
@@ -3208,4 +3209,36 @@ public void testBucketLogging() throws ExecutionException, InterruptedException
       RemoteStorageHelper.forceDelete(storage, loggingBucket, 5, TimeUnit.SECONDS);
     }
   }
+
+  @Test
+  public void testDeleteLifecycleRules() throws ExecutionException, InterruptedException {
```

Comment:
Following up, did you try deleting a single lifecycle rule?

Reply:
@frankyn 

I have tried to find the feasible way to delete a single lifecycle rule but its look like there is no possible way to do that and also found that across the all languages have same behavior which we have currently implemented.

**Suggestion** 
-  I think we should have to update the method name like `clearLifecycleRules `or `disableLifecycleRules `instead of `deleteLifecycleRules `to be more clear, `deleteLifecycleRules `might confused to user.

- we can also place a  `Note` there like delete individual rules through the console currently  this library have limited support  to delete single rule.

Reply:
Hi @athakor, apologies for the delay.
```java

    String lifecycleTestBucketName = RemoteStorageHelper.generateBucketName();
    storage.create(
            BucketInfo.newBuilder(lifecycleTestBucketName)
                    .setLocation("us")
                    .setLifecycleRules(
                            ImmutableList.of(
                                    new LifecycleRule(
                                            LifecycleAction.newSetStorageClassAction(StorageClass.COLDLINE),
                                            LifecycleCondition.newBuilder()
                                                    .setAge(1)
                                                    .setNumberOfNewerVersions(3)
                                                    .setIsLive(false)
                                                    .setCreatedBefore(new DateTime(System.currentTimeMillis()))
                                                    .setMatchesStorageClass(ImmutableList.of(StorageClass.COLDLINE))
                                                    .build()),
                                    new LifecycleRule(
                                            LifecycleAction.newDeleteAction(),
                                            LifecycleCondition.newBuilder()
                                                    .setAge(1)
                                                    .build()
                                    )))
                    .build());
    // Delete OLM rule.
    Bucket remoteBucket =
            storage.get(lifecycleTestBucketName, Storage.BucketGetOption.fields(BucketField.LIFECYCLE));
    int priorSize = remoteBucket.getLifecycleRules().size();
    ArrayList<LifecycleRule> lifecycleRules = new ArrayList(remoteBucket.getLifecycleRules());
    Iterator<LifecycleRule> iterator = lifecycleRules.iterator();
    while(iterator.hasNext()) {
      LifecycleRule rule = iterator.next();
      if (rule.getAction().getActionType().equals(LifecycleRule.DeleteLifecycleAction.TYPE)) {
        iterator.remove();
      }
    }
    remoteBucket.toBuilder().setLifecycleRules(lifecycleRules).build().update();
```

Reply:
Did this help? I think we can merge this change once you add a helper to Storage interface but no change is required for StorageRpc.java.

Reply:
@frankyn thanks for info, I tried your sample code but its not working please check the below response.

```
[
  LifecycleRule{
    lifecycleAction=DeleteLifecycleAction{
      actionType=Delete
    },
    lifecycleCondition=LifecycleCondition{
      age=1,
      createBefore=null,
      numberofNewerVersions=null,
      isLive=null,
      matchesStorageClass=null
    }
  },
  LifecycleRule{
    lifecycleAction=SetStorageClassLifecycleAction{
      actionType=SetStorageClass,
      storageClass=COLDLINE
    },
    lifecycleCondition=LifecycleCondition{
      age=1,
      createBefore=2020-05-15,
      numberofNewerVersions=3,
      isLive=false,
      matchesStorageClass=[
        COLDLINE
      ]
    }
  }
]
```
does it works on your end?

Reply:
Good catch, looks like there's an issue with updating LifecycleRules when the metadata information is set.

Here's a workaround for the bug:

```java
package com.example.storage;

// [START storage_quickstart]
// Imports the Google Cloud client library
import com.google.api.client.util.DateTime;
import com.google.cloud.storage.Bucket;
import com.google.cloud.storage.BucketInfo;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageClass;
import com.google.cloud.storage.StorageOptions;
import com.google.common.collect.ImmutableList;

import java.util.ArrayList;
import java.util.Iterator;

public class QuickstartSample {
  public static void main(String... args) throws Exception {
    // Instantiates a client
    Storage storage = StorageOptions.getDefaultInstance().getService();

    // The name for the new bucket
    String bucketName = "anima-frank-lifecycletestsdf";  // "my-new-bucket";

    // Creates the new bucket
    Bucket bucket = storage.create(BucketInfo.newBuilder(bucketName)
            .setLocation("us")
            .setLifecycleRules(
                    ImmutableList.of(
                            new BucketInfo.LifecycleRule(
                                    BucketInfo.LifecycleRule.LifecycleAction.newSetStorageClassAction(StorageClass.COLDLINE),
                                    BucketInfo.LifecycleRule.LifecycleCondition.newBuilder()
                                            .setAge(1)
                                            .setNumberOfNewerVersions(3)
                                            .setIsLive(false)
                                            .setCreatedBefore(new DateTime(System.currentTimeMillis()))
                                            .setMatchesStorageClass(ImmutableList.of(StorageClass.COLDLINE))
                                            .build()),
                            new BucketInfo.LifecycleRule(
                                    BucketInfo.LifecycleRule.LifecycleAction.newDeleteAction(),
                                    BucketInfo.LifecycleRule.LifecycleCondition.newBuilder()
                                            .setAge(1)
                                            .build()
                            )))
            .build());
    // Delete OLM rule.
    Bucket remoteBucket =
            storage.get(bucketName, Storage.BucketGetOption.fields(Storage.BucketField.LIFECYCLE));
    int priorSize = remoteBucket.getLifecycleRules().size();
    ArrayList<BucketInfo.LifecycleRule> lifecycleRules = new ArrayList(remoteBucket.getLifecycleRules());
    Iterator<BucketInfo.LifecycleRule> iterator = lifecycleRules.iterator();
    while(iterator.hasNext()) {
      BucketInfo.LifecycleRule rule = iterator.next();
      if (rule.getAction().getActionType().equals(BucketInfo.LifecycleRule.DeleteLifecycleAction.TYPE)) {
        iterator.remove();
        System.out.println("???");
      }
    }
    System.out.printf("Updated size: %d %n", lifecycleRules.size());
    System.out.println(lifecycleRules);
    remoteBucket =
            storage.get(bucketName, Storage.BucketGetOption.fields(Storage.BucketField.STORAGE_CLASS));

    remoteBucket = remoteBucket.toBuilder().setLifecycleRules(lifecycleRules).build().update();
    System.out.println(remoteBucket.getLifecycleRules());
    storage.delete(bucketName);
    System.out.printf("Bucket %s (updated: %d, prior: %d) created.%n", bucket.getName(), remoteBucket.getLifecycleRules().size(), priorSize);
  }
}
```

Reply:
@frankyn sorry for the delay and thanks for the workaround. Partially this will works in case of `empty bucket` but what will happen if `bucket `contain list of `blobs`? in that case it will gets failed with following error :
```javascript
{
  "code" : 409,
  "errors" : [ {
    "domain" : "global",
    "message" : "The bucket you tried to delete was not empty.",
    "reason" : "conflict"
  } ],
  "message" : "The bucket you tried to delete was not empty."
}
```
The main problem with this solution is that, it will remove whole `bucket `and that never get back. so i think we need to go with current implementation or we need to find something new for this.

Reply:
`storage.delete(bucketName)` is clean up for the example and not required by remove lifecycle rule. 

Please let me know if there is still confusion with the workaround. In short, the library should not require the workaround.

Reply:
@frankyn thanks for the clarification. It's works i will raise separate PR by adding these helper to Storage interface.  

Thank you for your help.

Reply:
Thank you! I appreciate your patience.

Reply:
@athakor I may have may confused, you're going to close this PR and make another right?